sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceDetection.scala

```diff
@@ -0,0 +1,127 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources
+
+import scala.util.Try
+
+import org.apache.commons.io.FilenameUtils
+import org.apache.hadoop.fs.{FileStatus, Path}
+import org.apache.hadoop.mapred.{FileInputFormat, JobConf}
+
+import org.apache.spark.{Logging, SparkException}
+import org.apache.spark.sql.SQLContext
+
+object DataSourceDetection extends Logging {
```

Comment:
I'm going to take another look at this later, but can you make sure the performance is ok when querying a large number of files on s3? i.e. ideally we should only need to read one file's metadata, rather than all files.